free5gc-operator/controllers/smfdeployment_controller.go

```diff
@@ -0,0 +1,648 @@
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"html/template"
+	"sort"
+	"time"
+
+	appsv1 "k8s.io/api/apps/v1"
+	apiv1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	resource "k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/go-logr/logr"
+	workloadv1alpha1 "github.com/nephio-project/api/nf_deployments/v1alpha1"
+)
+
+// SMFDeploymentReconciler reconciles a SMFDeployment object
+type SMFDeploymentReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+}
+
+type SMFcfgStruct struct {
+	PFCP_IP string // N4
+	// N4_IP   string
+	// N11_IP  string
+}
+
+type SMFAnnotation struct {
+	Name      string `json:"name"`
+	Interface string `json:"interface"`
+	IP        string `json:"ip"`
+	Gateway   string `json:"gateway"`
+}
+
+// Check string -> int/string -uint16 conversion
+func getSMFResourceParams(smfSpec workloadv1alpha1.SMFDeploymentSpec) (int32, *apiv1.ResourceRequirements, error) {
+	// Placeholder for Capacity calculation. Reqiurce requirements houlw be calculated based on DL, UL.
+
+	// TODO: increase number of replicas based on NFDeployment.Capacity.MaxSessions
+	var replicas int32 = 1
+	var cpuLimit string
+	var cpuRequest string
+	var memoryLimit string
+	var memoryRequest string
+
+	// spec.Capacity.MaxSessions <1000 -> 128Mi
+	// spec.Capacity.MaxNFConnections <10 -> 128Mi
+	if smfSpec.Capacity.MaxSessions < 1000 && smfSpec.Capacity.MaxNFConnections < 10 {
+		cpuLimit = "100m"
+		cpuRequest = "100m"
+		memoryRequest = "128Mi"
+		memoryLimit = "128Mi"
+	} else {
+		cpuLimit = "500m"
+		cpuRequest = "500m"
+		memoryRequest = "512Mi"
+		memoryLimit = "512Mi"
+	}
+
+	resources := apiv1.ResourceRequirements{}
+	resources.Limits = make(apiv1.ResourceList)
+	resources.Limits[apiv1.ResourceCPU] = resource.MustParse(cpuLimit)
+	resources.Limits[apiv1.ResourceMemory] = resource.MustParse(memoryLimit)
+
+	resources.Requests = make(apiv1.ResourceList)
+	resources.Requests[apiv1.ResourceCPU] = resource.MustParse(cpuRequest)
+	resources.Requests[apiv1.ResourceMemory] = resource.MustParse(memoryRequest)
+
+	return replicas, &resources, nil
+}
+
+func constructSMFNadName(templateName string, suffix string) string {
+	return templateName + "-" + suffix
+	// return "n4"
+}
+
+// getSMFNad returns NAD label string composed based on the Nx interfaces configuration provided in SMFDeploymentSpec
+func getSMFNad(templateName string, spec *workloadv1alpha1.SMFDeploymentSpec) string {
+	var ret string
+
+	// n6CfgSlice := getIntConfigSlice(spec.Interfaces, "N6")
```

Comment:
are these commented out things not needed or is this just temporary and will be updated in later PRs?